### PR TITLE
feat(orchestration): introducing the scheduler

### PR DESCRIPTION
**Pull request by TBonnin**
This is the first stage of the orchestration v2, implementing the scheduler.
As agreed the first version of the scheduler doesn't support recurring tasks, only tasks that are scheduled immediately.

`scheduler` package is not used yet but it is big enough to warrant a review. There is still a few TODOs, which I will address in subsequent PRs

## Issue ticket number and link
https://linear.app/nango/document/[brief]-orchestration-v2-4c5d1f1e104b#94b7a156-de22-4ce2-8a77-9cfcfe1fcece

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [x] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

File: package.json
Comment (bodinsamuel) [inline package.json:14]: missing in ./Dockerfile and tsconfig.docker.json  (we need to find a way to detect/automate this)

File: packages/scheduler/lib/db/migrations/20240506105059_initial_scheduler_models.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/db/migrations/20240506105059_initial_scheduler_models.ts]:
```suggestion
    await knex.raw(`DROP TABLE IF EXISTS "${TASKS_TABLE}"`);
```

File: packages/scheduler/lib/db/migrations/20240506105059_initial_scheduler_models.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/db/migrations/20240506105059_initial_scheduler_models.ts]: any reason to use jsonb and not just json?

File: packages/scheduler/package.json
Comment (bodinsamuel) [inline packages/scheduler/package.json]:
```suggestion
        "uuidv7": "0.6.3"
```

File: packages/scheduler/lib/vitest.d.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/vitest.d.ts]: Since I synced the vitest package version you should be able to point this in the tsconfig instead of copying

File: packages/scheduler/lib/models/tasks.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/models/tasks.ts:43]: maybe belong in types.ts?

File: packages/scheduler/lib/monitor.worker.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/monitor.worker.ts]: Suggestion to use native timers

```ts
import setTimeout from 'node:timers/promises'

await setTimeout(this.tickIntervalMs);
```

File: packages/scheduler/lib/types.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/types.ts]: I'll be careful of exporting const inside a types file, this can produce side effects when importing in other stuff

File: packages/scheduler/lib/monitor.worker.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/monitor.worker.ts]: stop (or another function) should clean up the listener, the ref and terminate otherwise it will block the graceful termination of the process.

File: packages/scheduler/lib/monitor.worker.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/monitor.worker.ts:94]: that why I'm not a fan of this way of doing things in nodejs, looking at `expiresIfTimeout` you are already logging an error so it will double log

File: packages/scheduler/lib/monitor.worker.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/monitor.worker.ts]:
```suggestion
            this.worker = new Worker(url);
```

File: packages/scheduler/lib/monitor.worker.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/monitor.worker.ts:98]:
```suggestion
                await this.send({ ids: taskIds });
```

Comment (bodinsamuel) [reply packages/scheduler/lib/monitor.worker.ts:98]: or remove async from send*

File: packages/scheduler/lib/scheduler.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/scheduler.ts:20]: I would be very cautious about jsdoc, unless we add an eslint package to enforce their consistency it's not going to be maintained properly

File: packages/scheduler/lib/scheduler.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/scheduler.ts:35]: should throw if this if is relevant no?

File: packages/scheduler/lib/scheduler.integration.test.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/scheduler.integration.test.ts]:
```suggestion
    it('should retry failed task if max retries is not reached', async () => {
```

**Review by bodinsamuel — COMMENTED** [review]
Massive 👏🏻 
Did not had time to test everything, but here is my first batch of comments

File: packages/scheduler/lib/models/tasks.integration.test.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/models/tasks.integration.test.ts]: we could add a special expect with the correct type like I did in shared/lib/vitest.d.ts

File: packages/scheduler/lib/models/tasks.integration.test.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/models/tasks.integration.test.ts:172]: every functions could be in test.helpers.ts for clarity

File: packages/scheduler/lib/models/tasks.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/models/tasks.ts]: could use get() here

File: packages/scheduler/lib/models/tasks.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/models/tasks.ts]: There is no global dequeue right now, how it will be managed? 
One while loop per groupKey?

--- 
old comment, nevermind I didn't pay attention of groupKey being passed as param

I would use `RANK OVER()`, 1 groupKey can fill the fullfill `limit`, bringing down every other customers which rank over could handle by default

File: packages/scheduler/lib/models/tasks.ts
Comment (bodinsamuel) [inline packages/scheduler/lib/models/tasks.ts]: Any reason to use a transaction here? you are doing everything in one query

**Review by bodinsamuel — COMMENTED** [review]
Finished reading, I'll let you reply and adjust what you feel relevant. Ping me when you want a second review ☺️

File: package.json
Comment (TBonnin) [reply package.json:14]: yep. I always forgot those

Comment (TBonnin) [reply package.json:14]: yep. Always forgot those

File: packages/scheduler/lib/db/migrations/20240506105059_initial_scheduler_models.ts
Comment (TBonnin) [reply packages/scheduler/lib/db/migrations/20240506105059_initial_scheduler_models.ts]: no strong reason behind this choice. `json` could do as well, I don't think we are gonna need to query or update the json content. I am happy to switch to `json` if you feel strongly about it

File: packages/scheduler/lib/scheduler.ts
Comment (TBonnin) [reply packages/scheduler/lib/scheduler.ts:20]: It is a risk indeed. Would you prefer no jsdoc?

File: packages/scheduler/lib/scheduler.ts
Comment (TBonnin) [reply packages/scheduler/lib/scheduler.ts:35]: yes good idea

File: packages/scheduler/lib/db/migrations/20240506105059_initial_scheduler_models.ts
Comment (bodinsamuel) [reply packages/scheduler/lib/db/migrations/20240506105059_initial_scheduler_models.ts]: not strongly, it just takes more space than necessary

File: packages/scheduler/lib/models/tasks.ts
Comment (TBonnin) [reply packages/scheduler/lib/models/tasks.ts:43]: it is purely internal and should not be expose outside of this file I think.

File: packages/scheduler/lib/types.ts
Comment (TBonnin) [reply packages/scheduler/lib/types.ts]: actually I don't want `taskStates` to be exposed outside of this package. Moved it to tasks.ts

File: packages/scheduler/lib/monitor.worker.ts
Comment (TBonnin) [reply packages/scheduler/lib/monitor.worker.ts:94]: not sure I follow. There is no logging in `expiresIfTimeout` or in any `tasks` function, errors are always returned

File: packages/scheduler/lib/models/tasks.ts
Comment (bodinsamuel) [reply packages/scheduler/lib/models/tasks.ts]: Discussed, for the moment it will be a global groupKey so no issue

File: packages/scheduler/lib/monitor.worker.ts
Comment (bodinsamuel) [reply packages/scheduler/lib/monitor.worker.ts:94]: yep, misread completely my bad

File: packages/scheduler/lib/models/tasks.integration.test.ts
Comment (TBonnin) [reply packages/scheduler/lib/models/tasks.integration.test.ts]: moved `vitest.d.ts` to utils and done what you mentioned

File: packages/scheduler/lib/models/tasks.integration.test.ts
Comment (TBonnin) [reply packages/scheduler/lib/models/tasks.integration.test.ts:172]: I'll move then if I ever need them somewhere else

File: packages/scheduler/lib/scheduler.ts
Comment (TBonnin) [reply packages/scheduler/lib/scheduler.ts:20]: I am keeping the comments for now.

File: packages/scheduler/lib/db/migrations/20240506105059_initial_scheduler_models.ts
Comment (TBonnin) [reply packages/scheduler/lib/db/migrations/20240506105059_initial_scheduler_models.ts]: changed to `json`

**Comment by TBonnin** [comment]
I addressed all your comments @bodinsamuel

**Review by bodinsamuel — APPROVED** [review]
Let's go 🚀

File: packages/scheduler/lib/models/tasks.ts
Comment (khaliqgant) [inline packages/scheduler/lib/models/tasks.ts:236]: Intentional I presume to leave this in while under development @TBonnin ?

**Comment by khaliqgant** [comment]
Looks good! Tests 💯

File: packages/scheduler/lib/models/tasks.ts
Comment (TBonnin) [reply packages/scheduler/lib/models/tasks.ts:236]: no actually it is some dev leftover. I will remove it

**Pull request merged**
